Add static assertions in generated serialization code to verify correct member ordering

achristensen07 opened the pull request:
#### b369d6af63fe770df55f19584fbd8dea19c43b40
<pre>
Add static assertions in generated serialization code to verify correct member ordering
<a href="https://bugs.webkit.org/show_bug.cgi?id=251536">https://bugs.webkit.org/show_bug.cgi?id=251536</a>
rdar://104925697

Reviewed by Ben Nham.

Also document all attributes and remove unneeded ones.

* Source/WebCore/Modules/indexeddb/shared/IDBDatabaseInfo.cpp:
(WebCore::IDBDatabaseInfo::IDBDatabaseInfo):
* Source/WebCore/Modules/indexeddb/shared/IDBDatabaseInfo.h:
* Source/WebCore/platform/graphics/Pattern.h:
* Source/WebKit/Scripts/generate-serializers.py:
(SerializedType.__init__):
(SerializedType.can_assert_member_order_is_correct):
(check_type_members):
* Source/WebKit/Scripts/webkit/tests/GeneratedSerializers.cpp:
(IPC::ArgumentCoder&lt;Namespace::OtherClass&gt;::encode):
(IPC::ArgumentCoder&lt;Namespace::EmptyConstructorNullable&gt;::encode):
(IPC::ArgumentCoder&lt;NullableSoftLinkedMember&gt;::encode):
* Source/WebKit/Scripts/webkit/tests/TestSerializedType.serialization.in:
* Source/WebKit/Shared/API/APIError.serialization.in:
* Source/WebKit/Shared/API/APIFrameHandle.serialization.in:
* Source/WebKit/Shared/API/APIGeometry.serialization.in:
* Source/WebKit/Shared/API/APIPageHandle.serialization.in:
* Source/WebKit/Shared/API/APIURL.serialization.in:
* Source/WebKit/Shared/API/APIURLRequest.serialization.in:
* Source/WebKit/Shared/API/APIURLResponse.serialization.in:
* Source/WebKit/Shared/SessionState.serialization.in:
* Source/WebKit/Shared/ShareableBitmap.serialization.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/Shared/mac/SecItemResponseData.h:
(WebKit::SecItemResponseData::SecItemResponseData):

Canonical link: <a href="https://commits.webkit.org/259784@main">https://commits.webkit.org/259784@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/be856658fc31583f53e0b23a2bcb4e880fe8599a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/105807 "15 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/14875 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/38655 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/115005 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/175132 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/109709 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/16294 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/6074 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/98036 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/114791 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/111558 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/12393 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/95367 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/39851 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/5/builds/109253 "Passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/94257 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/27008 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/81581 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/8146 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/28360 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/8628 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/4946 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/14255 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/47908 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/10185 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/3630 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

bnham commented on Source/WebKit/Scripts/generate-serializers.py:
```diff
@@ -25,6 +25,22 @@
 import re
 import sys
 
+# Supported type attributes:
```
👍

bnham commented on Source/WebKit/Scripts/generate-serializers.py:
```diff
@@ -469,6 +509,17 @@ def generate_impl(serialized_types, serialized_enums, headers):
+    result.append('template<size_t...> struct MembersInCorrectOrder;')
+    result.append('template<size_t onlyOffset> struct MembersInCorrectOrder<onlyOffset> { static constexpr bool value = true; };')
+    result.append('template<size_t firstOffset, size_t secondOffset, size_t... remainingOffsets> struct MembersInCorrectOrder<firstOffset, secondOffset, remainingOffsets...> {')
+    result.append('    static constexpr bool value = firstOffset > secondOffset ? false : MembersInCorrectOrder<secondOffset, remainingOffsets...>::value;')
```
Discussed this offline with Alex and we are using `>` instead of `>=` here to support things like serializing zero-sized members, e.g. those marked with `[[no_unique_address]]`.

bnham reviewed (approved):
LGTM

achristensen07 committed:
Add static assertions in generated serialization code to verify correct member ordering
https://bugs.webkit.org/show_bug.cgi?id=251536
rdar://104925697

Reviewed by Ben Nham.

Also document all attributes and remove unneeded ones.

* Source/WebCore/Modules/indexeddb/shared/IDBDatabaseInfo.cpp:
(WebCore::IDBDatabaseInfo::IDBDatabaseInfo):
* Source/WebCore/Modules/indexeddb/shared/IDBDatabaseInfo.h:
* Source/WebCore/platform/graphics/Pattern.h:
* Source/WebKit/Scripts/generate-serializers.py:
(SerializedType.__init__):
(SerializedType.can_assert_member_order_is_correct):
(check_type_members):
* Source/WebKit/Scripts/webkit/tests/GeneratedSerializers.cpp:
(IPC::ArgumentCoder<Namespace::OtherClass>::encode):
(IPC::ArgumentCoder<Namespace::EmptyConstructorNullable>::encode):
(IPC::ArgumentCoder<NullableSoftLinkedMember>::encode):
* Source/WebKit/Scripts/webkit/tests/TestSerializedType.serialization.in:
* Source/WebKit/Shared/API/APIError.serialization.in:
* Source/WebKit/Shared/API/APIFrameHandle.serialization.in:
* Source/WebKit/Shared/API/APIGeometry.serialization.in:
* Source/WebKit/Shared/API/APIPageHandle.serialization.in:
* Source/WebKit/Shared/API/APIURL.serialization.in:
* Source/WebKit/Shared/API/APIURLRequest.serialization.in:
* Source/WebKit/Shared/API/APIURLResponse.serialization.in:
* Source/WebKit/Shared/SessionState.serialization.in:
* Source/WebKit/Shared/ShareableBitmap.serialization.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/Shared/mac/SecItemResponseData.h:
(WebKit::SecItemResponseData::SecItemResponseData):

Canonical link: https://commits.webkit.org/259784@main
